plutus-core/untyped-plutus-core/src/UntypedPlutusCore/Evaluation/Machine/Cek/StepCounter.hs

```diff
@@ -0,0 +1,97 @@
+{-# LANGUAGE BangPatterns #-}
+module UntypedPlutusCore.Evaluation.Machine.Cek.StepCounter where
+
+import Control.Monad.Primitive
+import Data.Primitive qualified as P
+import Data.Word
+
+-- See Note [Step counter data structure]
+-- | A set of 'Word8' counters that is used in the CEK machine
+-- to count steps.
+newtype StepCounter s = StepCounter (P.MutablePrimArray s Word8)
+
+{-# INLINE newCounter #-}
+-- | Make a new 'StepCounter' with the given number of counters.
+newCounter :: PrimMonad m => Int -> m (StepCounter (PrimState m))
+newCounter sz = do
+  c <- StepCounter <$> P.newPrimArray sz
+  resetCounter c
+  pure c
+
+{-# INLINE resetCounter #-}
+-- | Reset all the counters in the given 'StepCounter' to zero.
+resetCounter :: PrimMonad m => StepCounter (PrimState m) -> m ()
+resetCounter (StepCounter arr) = P.setPrimArray arr 0 (P.sizeofMutablePrimArray arr) 0
+
+{-# INLINE readCounter #-}
+-- | Read the value of a counter.
+readCounter :: PrimMonad m => StepCounter (PrimState m) -> Int -> m Word8
+readCounter (StepCounter arr) = P.readPrimArray arr
```

Comment:
It doesn't work without (perhaps) lots of type annotations. I'm fairly confident in the effect of it being newtypes + INLINE!

Reply:
I think it's not quite the same though. An excerpt from `Data.Functor.Utils` from `base`:

```
-- See Note [Function coercion]
(#.) :: Coercible b c => (b -> c) -> (a -> b) -> (a -> c)
(#.) _f = coerce
{-# INLINE (#.) #-}

{-
Note [Function coercion]
~~~~~~~~~~~~~~~~~~~~~~~

Several functions here use (#.) instead of (.) to avoid potential efficiency
problems relating to #7542. The problem, in a nutshell:

If N is a newtype constructor, then N x will always have the same
representation as x (something similar applies for a newtype deconstructor).
However, if f is a function,

N . f = \x -> N (f x)

This looks almost the same as f, but the eta expansion lifts it--the lhs could
be _|_, but the rhs never is. This can lead to very inefficient code.  Thus we
steal a technique from Shachaf and Edward Kmett and adapt it to the current
(rather clean) setting. Instead of using  N . f,  we use  N #. f, which is
just

coerce f `asTypeOf` (N . f)

That is, we just *pretend* that f has the right type, and thanks to the safety
of coerce, the type checker guarantees that nothing really goes wrong. We still
have to be a bit careful, though: remember that #. completely ignores the
*value* of its left operand.
-}
```

In case of `readCounter` you _somewhat_ require the argument to be provided like is explained in

```haskell
{-# INLINE (.) #-}
-- Make sure it has TWO args only on the left, so that it inlines
-- when applied to two functions, even if there is no final argument
(.)    :: (b -> c) -> (a -> b) -> a -> c
(.) f g = \x -> f (g x)
```


Now chances are, `readCounter` will get inlined regardless, but you'll end up with ``\arr ->  P.readPrimArray (arr `cast` <...>)`` instead of ``P.readPrimArray `cast` <...>``. Having to write types is very common when using `coerce` (which is how we have `coerceVia`, `coerceArg` and `coerceRes` in `PlutusPrelude` to minimize that), but I believe it's worth it in places where we care about performance.